### PR TITLE
Settings: NVS: Use hash as vector offset

### PR DESCRIPTION
**Pull request by LingaoM**
Use hash as vector offset to speed up search and save some memory.
For conflicting name ids, only the maximum value is taken. In this way, the subsequent traversal logic can be used and redundant nvs_read can be reduced.

### CODE REVIEW

**Comment by LingaoM** [comment]
@de-nordic @Laczen @Damian-Nordic CC

**Comment by Damian-Nordic** [comment]
@LingaoM I think that this implementation may not work correctly in all cases. Consider the following scenario:
1. There are two keys stored in the settings, _key1_ and _key2_, such that _hash(key1) == hash(key2)_ and _nvs_id(key1) < nvs_id(key2)_.
2. The device is rebooted but none of the keys are loaded.
3. _key1_ is saved. As a result, the cache is populated with _nvs_id(key1)_.
4. _key2_ is about to saved. Due to the content of the cache, the key is looked up from _nvs_id(key1)_ in descending order. As a result, _key2_ is stored again using potentially already reserved NVS ID.

In other words, I think that if you store `max` in the cache, you must be absolutely sure that all keys are taken into account. Therefore, you probably would need to populate the cache at the settings initalization. Also, I think that `write_name_id` should not be initialized to the return value of `settings_nvs_cache_match()` - consider another scenario, in which you save two keys _key1_ and _key2_, such that _hash(key1) == hash(key2)_. When saving _key2_, `settings_nvs_cache_match()` may return a match, but _key2_ is not found in the NVS, so it should be written to `cf->last_name_id + 1` instead of the cached NVS ID + 1.

**Comment by LingaoM** [comment]
I want to say that for NVS, if the load is not called when the power is turned on, but only the save operation is performed, although this is allowed, I can't think of the meaning of this existence. The point you pointed out still has significance. Thank you very much. I think without Turn off the cache function by performing a save operation after loading. This will solve this problem.

**Comment by Damian-Nordic** [comment]
@LingaoM I agree that first scenario is not something common, but still possible. But the second scenario I described is pretty likely to happen and it's still not addressed in the newest patch, so please take a look at that.

**Comment by Laczen** [comment]
@LingaoM I also have my doubts about the method working correctly in all cases. Is it possible to create a test that evaluates what happens when 2 items with the same hash are stored: when the newest item is deleted will the other one still be found ?

**Comment by LingaoM** [comment]
@Damian-Nordic @Laczen I'm enable cache in the test configuration file, and ci passed, please take a look again, thanks.

Due to no more time currently, more test case will todo in the future, not in this pull request.

File: subsys/settings/src/settings_nvs.c [hunk 1]
```diff
@@ -183,6 +164,10 @@ static int settings_nvs_load(struct settings_store *cs,
 			break;
 		}
 	}
+
+#if CONFIG_SETTINGS_NVS_NAME_CACHE
+	cf->loaded = true;
```
Comment (Damian-Nordic) [inline subsys/settings/src/settings_nvs.c:169]: Note that `settings_call_set_handler` may return 1 (for instance, the user callback may decide not to continue loading more keys), in which case the cache may still not be fully populated. Why not to populate the cache at the initialization to be sure that all keys are iterated over?

**Comment by Laczen** [comment]
> @Damian-Nordic @Laczen I'm enable cache in the test configuration file, and ci passed, please take a look again, thanks.
> 
> Due to no more time currently, more test case will todo in the future, not in this pull request.

@LingaoM, correct me if I'm wrong but the hash table you are using is only meant to speed up writing of settings. If that is correct what if you would do it in the following way:
1. You intialize the hash table with a invalid id: a good choice would be NVS_NAMECNT_ID + NVS_NAME_ID_OFFSET,
2. You introduce a variable for the lowest name_id that is covered by the hash table: hash_table_low_id, that is initialized to `cf->last_name_id`,
3. When you do a write you lookup the name_id and when it is found you use that name_id, when it is invalid you can set the name_id to the hash_table_low_id,
4. When you do a read of the name_id (and it is not the hash_table_low_id) and there is no entry you can make a mark that that the id in the hash_table can be updated,
5. Everytime you reduce the name_id you can also reduce the hash_table_low_id (until it reaches NVS_NAMECNT_ID),
6. If hash_table_low_id is at NVS_NAMECNT_ID and there is no match in the hash table it is new data and `cf->last_name_id` needs to be increased and a new entry needs to be made in the hash table.

**Pull request closed without merging**